flow/flow_ebos_blackoil_tpfa.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "config.h"
+
+#include <flow/flow_ebos_blackoil_tpfa.hpp>
+
+#include <opm/material/common/ResetLocale.hpp>
+#include <opm/grid/CpGrid.hpp>
+#include <opm/simulators/flow/SimulatorFullyImplicitBlackoilEbos.hpp>
+#include <opm/simulators/flow/Main.hpp>
+
+#include <opm/models/blackoil/blackoillocalresidualtpfa.hh>
+#include <opm/models/discretization/common/tpfalinearizer.hh>
+
+namespace Opm {
+    namespace Properties {
+        namespace TTag {
+            struct EclFlowProblemTPFA {
+            using InheritsFrom = std::tuple<EclFlowProblem>;
+            };
+        }
+   }
+}
+
+namespace Opm {
+    namespace Properties {
+
+        template<class TypeTag>
+        struct Linearizer<TypeTag, TTag::EclFlowProblemTPFA> { using type = TpfaLinearizer<TypeTag>; };
+
+        template<class TypeTag>
+        struct LocalResidual<TypeTag, TTag::EclFlowProblemTPFA> { using type = BlackOilLocalResidualTPFA<TypeTag>; };
+
+        template<class TypeTag>
+        struct EnableDiffusion<TypeTag, TTag::EclFlowProblemTPFA> { static constexpr bool value = false; };
+
```

Comment:
Is this for further speed-up or are there unexpected changes that disallow diffusion in combination with these changes.

Reply:
While it may or may not give a speedup (I think you reported no significant cost at the time the default was changed to true), the reason here is that the flux and/or source contributions for most modules apart from blackoil have not been refactored yet. For some it will be quick and easy, for those that have shear-dependent (i.e. flux dependent) behaviour it will be a bit more work.

You can see in the opm-models PR, in blackoilresidualtpfa.hh lines 368-398. Those guarantee for now that we do not make incompatible combinations that would silently fail. My goal is that we attack them one by one as we find time and need to do it.